### PR TITLE
Added frame.waitFor(selector)

### PR DESCRIPTION
**Pull request by alexkozy**
Bug: https://github.com/GoogleChrome/puppeteer/issues/42

### CODE REVIEW

File: lib/FrameManager.js [hunk 1]
```diff
@@ -197,6 +197,33 @@ class FrameManager extends EventEmitter {
     });
     return response.result.value;
   }
+
+  /**
+   * @param {!Frame} frame
```
Comment (paulirish) [inline lib/FrameManager.js]: missing selector arg in jsdoc

File: lib/FrameManager.js [hunk 2]
```diff
@@ -197,6 +197,33 @@ class FrameManager extends EventEmitter {
...
+   * @return {!Promise<undefined>}
+   */
+  async _waitForInFrame(frame, selector) {
```
Comment (paulirish) [inline lib/FrameManager.js]: i'd swap the order of these two

Comment (alexkozy) [reply lib/FrameManager.js]: done.

File: lib/FrameManager.js (hunk 1 above)
Comment (alexkozy) [reply lib/FrameManager.js]: done.

File: lib/FrameManager.js [hunk 3]
```diff
@@ -197,6 +197,34 @@ class FrameManager extends EventEmitter {
...
   }
+
+  /**
+   * @param {string} selector
+   * @param {!Frame} frame
+   * @return {!Promise<undefined>}
+   */
+  async _waitForInFrame(selector, frame) {
+    let code = selector => new Promise((fulfill, reject) => {
+      new MutationObserver((mutations, observer) => {
```
Comment (aslushnikov) [inline lib/FrameManager.js]: let's check if it already exists

File: lib/FrameManager.js [hunk 4]
```diff
@@ -197,6 +197,34 @@ class FrameManager extends EventEmitter {
...
+          }
+        }
+      }).observe(document.documentElement, {
+        childList: true,
+        subtree: true
+      });
+    });
+    await this._client.send('Runtime.evaluate', {
+      expression: helper.evaluationString(code, selector),
+      awaitPromise: true
```
Comment (aslushnikov) [inline lib/FrameManager.js]: returnByValue: true

**Review by aslushnikov — CHANGES_REQUESTED** [review]

File: lib/FrameManager.js (hunk 4 above)
Comment (alexkozy) [reply lib/FrameManager.js]: done.

File: lib/FrameManager.js (hunk 3 above)
Comment (alexkozy) [reply lib/FrameManager.js]: done, test added!

**Pull request merged**